### PR TITLE
Debug output for bats tests and fix for Grafana installer

### PR DESCRIPTION
**Pull request by holgerfriedrich**
* add output for all failed tests
* add lsb-release, unzip to Docker image used for bats #917
* fixed influxdb+grafana installation to run properly in CI #917 #921

Signed-off-by: Holger Friedrich <mail@holger-friedrich.de>

### CODE REVIEW

**Comment by holgerfriedrich** [comment]
@mstormi getting closer, pls. take a look at the CI. Added output to all bats checks.

**Comment by mstormi** [comment]
> take a look at the CI

what do you mean?

**Comment by holgerfriedrich** [comment]
https://travis-ci.org/github/holgerfriedrich/openhabian/jobs/692049537 (hope you can open this, the PR build is still running)
https://travis-ci.com/github/openhab/openhabian/jobs/340822671

**Comment by mstormi** [comment]
Ok but what doe you mean by your comment ? That you can now produce output in BATS tests to appear in CI logs and that we should look at your Grafana bats code to see how ?

**Comment by holgerfriedrich** [comment]
I added it to all tests
```
  if [ "$status" -ne 0 ]; then echo "$output" >&3; fi
  [ "$status" -eq 0 ]
```

Now you can see what the other tests are complaining about, e.g.
```
# Zulu 64-bit Java installation is being (test-)installed...

# Zulu 64-bit Java installation successful.

$ systemctl stop openhab2.service 

ERROR:systemctl:Unit openhab2.service could not be found.

/opt/openhabian/./functions/java-jre.bash: line 65: timestamp: command not found
```

But still I cannot get Grafan install to start up the service properly. Will have to look at it again in the evening...

**Comment by holgerfriedrich** [comment]
@ecdye any idea why it stall hangs after rebasing on #930?

**Comment by ecdye** [comment]
I have my theories but I'll wait to make any statements until I have confirmed them on my end.

**Comment by ecdye** [comment]
Ok, so I think it has to do with adding the packages directly to the dockerfiles. Rather install the packages in the script that you are testing. For example, instead of adding `software-properties-common` directly to the Dockerfile (as when I did this was when Bats would hang, don't ask why because I have no clue) I added this to my install script for AdoptOpenJDK (what required the `software-properties-common` package):
``` bash
if ! dpkg -s 'software-properties-common' > /dev/null 2>&1; then
    cond_redirect apt-get install --yes software-properties-common
fi
```
Then after discovering the error mentioned below relating to influx-db through the use of your debug command (very useful), and working around it, my latest build worked correctly without any hanging. See <https://travis-ci.com/github/openhab/openhabian/jobs/341218969> for the results.

As an aside, it appears that influx-db install is not correct for the `apt` lists as I had to work around it and that was why my Adopt test was failing. See <https://travis-ci.com/github/openhab/openhabian/jobs/341215297#L459> for what I mean. I have no idea if you already fixed this here but I just used a setup script in bats to clear the apt lists directory before running the Java tests and that solved the issue.

That is what worked for me, however, your mileage may vary. Let me know if what I said makes no sense and I can try to clarify.

**Comment by mstormi** [comment]
> Then after discovering the error mentioned below relating to influx-db through the use of your debug command (very useful), and working around it, my latest build worked correctly without any hanging.

Then your builds were probably hung because of influx ? And you also fixed some ";" in influx install which made that going.
That now sounds more reasonable to be a reason than your move of a single package installation from Docker into the code.

**Comment by ecdye** [comment]
Sort of APT was complaining about being unable to read the file relating infulx so I made sure it would be deleted when starting my install script to reduce the error sources.

**Comment by holgerfriedrich** [comment]
@ecdye can you show where?
@ecdye influx apt problem due to missing lsb-release package, then apt line will be generated wrong. will add a check for this later.

I did a bit testing: adding bats debug output did not make the tests hang. adding the missing packages (unzip, lsb-release) both did (also separately), so I suppose it is related to the content of the tests which are then able to proceed.

I am aware of https://github.com/bats-core/bats-core#file-descriptor-3-read-this-if-bats-hangs, but I do not see where this could help.
What is the thing about missing ;  ???

**Comment by ecdye** [comment]
Show where what? The apt fail I mentioned it earlier its here: https://travis-ci.com/github/openhab/openhabian/jobs/341215297#L459 

The workaround for bats I used is the setup function (bats provides this and it runs before the tests) in my case it is very simple
``` bash
setup() {
  rm -rf /etc/apt/sources.list.d/*
}
```
the ; missing was just a theory of mine as to why the tests were hanging but apparently it was wrong.

**Comment by mstormi** [comment]
will you make it by tomorrow ? Would love to have BATS tests show green when we announce java beta test tomorrow but don't hurry, reviews and PR fixes for tomorrow are more important if in question.

**Comment by mstormi** [comment]
closure intentional?

**Comment by holgerfriedrich** [comment]
@mstormi will get this done soon. Bats tests were hanging once the installer worked.
Seems to be caused by services started by the test. Now added a teardown function.
Trying to finalize.

**Comment by holgerfriedrich** [comment]
@mstormi what should we do about tests like this, which currently fail:
```
@test "installation-java_exist" {
  run java -version
  if [ "$status" -ne 0 ]; then echo "$output" >&3; fi
  [ "$status" -eq 0 ]
  [[ $output == *"Zulu"* ]]
}
```
Current output:
```
/usr/local/lib/bats-core/test_functions.bash: line 38: java: command not found
not ok 1 installation-java_exist
```

**Comment by mstormi** [comment]
that particular one should be removed, does not make sense anyway

**Comment by mstormi** [comment]
Fix frontail test while you're at it.
I consider this to be enough. Else to have the service running we would need java installed

[22:12:37] openhabian@openhabianpi:~/openhabian$ cat functions/nodejs-apps.bats
#!/usr/bin/env bats

load nodejs-apps
load helpers

@test "installation-Frontail_is_running" {
  run frontail_setup
  run frontail -V
  if [ "$status" -ne 0 ]; then echo "$output" >&3; fi
  [ "$status" -eq 0 ]

  echo -e "# \e[32mFrontail properly installed." >&3
}

**Comment by mstormi** [comment]
you said you install lsb-release now ? That's also needed for the homegear test

**Comment by holgerfriedrich** [comment]
frontail test fails, see https://travis-ci.org/github/holgerfriedrich/openhabian/jobs/693227967

and yes, need to install lsb-release.
currently trying to trace which test still causes bats to hang after completion.

**Comment by mstormi** [comment]
did you fix the InfluxDB repo ? For me that caused the next bats test in line to fail on apt-update

**Comment by holgerfriedrich** [comment]
Rebased, now all tests are hopefully green again.
Grafana install needs to be checked on real hardware and influxdb needs to be tested it it started after reboot.

**Comment by holgerfriedrich** [comment]
@mstormi finally rebased and ready for review.
Just purged influx and grafana on my test install, reinstall worked fine and influx still works after reboot.

File: Dockerfile.openhabian-aarch64-native [hunk 1]
```diff
@@ -9,7 +9,7 @@ ENV INITSYSTEM on
 ENV APT_KEY_DONT_WARN_ON_DANGEROUS_USAGE=DontWarn
 
 RUN apt-get update -qq \
-    && apt-get install --yes -qq --no-install-recommends systemd systemd-sysv git wget python3 python3-pip apt-utils jq virt-what \
+    && apt-get install --yes -qq --no-install-recommends systemd systemd-sysv git wget python3 python3-pip apt-utils jq virt-what lsb-release unzip psmisc \
```
Comment (mstormi) [inline Dockerfile.openhabian-aarch64-native:12]: Maybe create a separate Dockerfile for BATS, else we'll forget why this line is different across Dockerfiles.

File: functions/packages.bats [hunk 2]
```diff
@@ -3,12 +3,29 @@
...
   [ "$status" -eq 0 ]
   echo -e "# \e[32mHomegear installation successful." >&3
-  run systemctl is-active --quiet homegear.service
+  # TODO: fix homegear service, currently it does not detect running instance in CI
+  #run systemctl status homegear.service
+  #echo "$output" >&3
+  #run systemctl is-active --quiet homegear.service
+  #if [ "$status" -ne 0 ]; then echo "$output" >&3; fi
+  # workaround in place:
+  run bash -c "ps aux | grep homegear"
```
Comment (mstormi) [inline functions/packages.bats:28]: will always find the grep process so no good test

Comment (holgerfriedrich) [reply functions/packages.bats]: right, now fixed

File: Dockerfile.openhabian-aarch64-native (hunk 1 above)
Comment (holgerfriedrich) [reply Dockerfile.openhabian-aarch64-native:18]: Added a comment block, IMHO better than creating another config.
Maybe we should silently extend the installed packages also in all other config files.

**Pull request merged**